src/System.Text.Primitives/System/Text/Encoding/Utf8/Utf8Encoder.cs

```diff
@@ -290,53 +300,119 @@ internal static bool TryComputeEncodedBytes(ReadOnlySpan<uint> utf32, out int by
         /// <returns>True if the input buffer was fully encoded into the output buffer, otherwise false.</returns>
         public static bool TryEncode(ReadOnlySpan<char> utf16, Span<byte> utf8, out int charactersConsumed, out int bytesWritten)
         {
-            var availableCharacters = utf16.Length;
-            var bufferLength = utf8.Length;
+            var outputBytesAvailable = utf8.Length;
+            var inputCharacters = utf16.Length;
 
-            charactersConsumed = 0;
-            bytesWritten = 0;
+            int bytesWrittenTemp = 0;
+            int i = 0;
+            char codePoint = '\0';
 
-            for (int i = 0; i < availableCharacters; i++)
+            unsafe
             {
-                var ch = utf16[i];
+                var utf16Ref = Unsafe.AsPointer(ref utf16.DangerousGetPinnableReference());
```

Comment:
I likely means that your buggy implementation using AsPointer was not operating correctly - it might have been operating on corrupted data, hard to tell. You can make things run really-really fast when they do not have to work...